presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java

```diff
@@ -288,25 +282,107 @@ public void setWindowFunctions(QuerySpecification node, List<FunctionCall> funct
         return windowFunctions.get(query);
     }
 
-    public void setOutputDescriptor(RelationType descriptor)
+    public Scope getScope(Node node)
     {
-        outputDescriptor = descriptor;
+        return tryGetScope(node).orElseThrow(() -> new IllegalArgumentException(String.format("Analysis does not contain information for node: %s", node)));
     }
 
-    public RelationType getOutputDescriptor()
+    public Optional<Scope> tryGetScope(Node node)
+    {
+        // TODO check performance impact for planning of getScope method
+        if (scopes.containsKey(node)) {
+            return Optional.of(scopes.get(node));
+        }
+
+        Optional<Node> root = roots.stream()
+                .filter(i -> nodeContains(i, node))
+                .findAny();
+
+        if (root.isPresent()) {
+            GetScopeVisitor visitor = new GetScopeVisitor(scopes, node);
+            visitor.process(root.get(), new LinkedList<>());
+            return visitor.getResult();
+        }
+        return Optional.empty();
+    }
+
+    public Node getRoot()
```

Comment:
yeah... it was hacky. It come from the fact `StatementAnalyzer` were doing a query rewrites. I wanted to have one `root`, which I could traverse when looking for a `Scope`. But in case of a rewrite it become tricky. I could lose information about old statement `root` by overwriting the `root` or I could have multiple roots. Both solutions were bad.

I extracted rewrites from `StatementAnalyzer`. Now I have one `root`, `StatementAnalyzer` is more specialized (focused) and easier to understand.